modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -114,74 +128,125 @@ object GlobalSnapshotStateChannelEventsProcessor {
-                  case (state, Nil) => state.asRight[Agg].pure[F]
-
-                  case (None, head :: tail) =>
-                    jsonBrotliBinarySerializer
-                      .deserialize[Signed[CurrencySnapshot]](head.value.content)
-                      .map {
-                        _.toOption.map { snapshot =>
-                          (NonEmptyList.one(snapshot.asLeft).some, tail)
-                            .asLeft[Result]
-                        }
+      )(implicit hasher: Hasher[F]): F[SortedMap[Address, MetagraphAcceptanceResult]] = {
+        val isFeeRequired = feeCalculator.isFeeRequired(snapshotOrdinal)
+
+        def deserialize[A: Decoder](binary: Signed[StateChannelSnapshotBinary]): F[Option[A]] =
+          jsonBrotliBinarySerializer.deserialize[A](binary.value.content).map(_.toOption)
+
+        events.toList.foldLeftM(SortedMap.empty[Address, MetagraphAcceptanceResult]) {
+          case (agg, (address, binaries)) =>
+            type Result = Option[MetagraphAcceptanceResult]
+            type Agg = (Result, List[Signed[StateChannelSnapshotBinary]])
+
+            val stubBinary: Signed[StateChannelSnapshotBinary] = Signed(
+              StateChannelSnapshotBinary(Hash.empty, Array.emptyByteArray, SnapshotFee.MinValue),
+              NonEmptySet.one(SignatureProof(Id(Hex("")), Signature(Hex(""))))
+            )
+
+            val emptyBalanceUpdate = Map.empty[Address, Balance]
+
+            val initialState =
+              lastGlobalSnapshotInfo.lastCurrencySnapshots
+                .get(address)
+                .map(init => (stubBinary, init.some))
+                .map(s => (NonEmptyList.one(s), Map.empty[Address, Balance]))
+
+            (initialState, binaries.toList.reverse)
+              .tailRecM[F, Result] {
+                case (state, Nil) => state.asRight[Agg].pure[F]
+
+                case (None, head :: tail) =>
+                  deserialize[Signed[CurrencySnapshot]](head).map {
+                    case Some(snapshot) => // full snapshot - we don't subtract fee
+                      (
+                        (NonEmptyList.one((head, snapshot.asLeft.some)), emptyBalanceUpdate).some,
+                        tail
+                      ).asLeft
+                    case None => // no full snapshot yet - we only accept the binary if fee is not required
+                      if (isFeeRequired) none.asRight
+                      else ((NonEmptyList.one((head, none)), emptyBalanceUpdate).some, tail).asLeft
+                  }
+
+                case (current @ Some((nel, balanceUpdate)), head :: tail) =>
+                  nel.head match {
+                    case (_, None) =>
+                      deserialize[Signed[CurrencySnapshot]](head).map {
+                        case Some(snapshot) => // full snapshot - we don't subtract fee
+                          (
+                            (nel.prepend((head, snapshot.asLeft.some)), balanceUpdate).some,
```

Comment:
Is there a reason for using `prepend` in this class instead of `::` ?

Reply:
There is no difference between `prepend` and `::` right? Here anyway I think it's more readable to use `.prepend()` given that added object is more complex.